### PR TITLE
Add ability to create builds through API

### PR DESCRIPTION
**Pull request by claudioclutter**
Extracted from #167 to only focus on the API component,
adding tests and removing the changes to version and title.

Closes #167 -- to be rebased after #170

### CODE REVIEW

**Comment by claudioclutter** [comment]
Only check the last commit -- the other ones are there to build on top of #170

File: README.md [hunk 1]
```diff
@@ -52,6 +73,7 @@ The following environment variables are optional:
 
 - `GITHUB_PROJECT`: The GitHub "username/project" path to fetch release notes from
 - `SLACK_CHANNEL`: The Slack channel to post notifications to (defaults to #deploys)
+- `MDMA_TOKEN`: An authorization token for third-party APIclients
```
Comment (jeremygrenier) [inline README.md]: Typo in `APIclients` → `API clients`.

Comment (claudioclutter) [reply README.md]: Fixed! Thanks

**Comment by claudioclutter** [comment]
@jeremygrenier Apart from adding `MDMA_TOKEN` to existing app, I have to think about how to add to Review Apps.

The option I have in mind is to just add to app.json like other ENVs:

https://github.com/clutter/mdma/blob/5636310dd94946c475f2040afca175318ac424d7/app.json#L6-L10

So we hard-code a token that only works for staging (Review Apps) machines.
What are your thoughts?

**Comment by claudioclutter** [comment]
Another comment… 
now that we are extracting the `bundle_version` from the package **before** saving the build, how do you feel about also extracting the bundle identifier and then **checking** that it matches `MDMA_APP_IDENTIFIER`?

In other words… if someone tries to upload a "valid" IPA file but it's for the right app (e.g. you try to upload google_maps.ipa to an instance of mdma dedicated to facebook_maps.ipa) then you get a validation error.

**Comment by jeremygrenier** [comment]
For me that make sense to have the `MDMA_TOKEN` hardcoded for staging (Review Apps) machines.

Concerning the comparison with `MDMA_APP_IDENTIFIER` it is also a good idea

**Comment by claudioclutter** [comment]
@jeremygrenier After thinking about it, I would probably go with https://devcenter.heroku.com/articles/app-json-schema#env

Every review app will have a different MDMA TOKEN.
If you want to test third-party API clients against a review app, you have to read it from `heroku config:get` first.

**Comment by claudioclutter** [comment]
@jeremygrenier Looks like the random approach worked! Try this command

`heroku config:get MDMA_TOKEN -a mdma-field-pr-171`

**Comment by claudioclutter** [comment]
Great job @jeremygrenier !!

You were able to create an internal build via API from a GitHub action!!

Screen Shot 2019-11-13 at 12 45 00 PM

Maybe in the future it might be worth adding a section to the README of this repo (or to the wiki) to explain people how they can do the same, that is, how do they have to write a GitHub action to take advantage of this.

**Comment by claudioclutter** [comment]
@jeremygrenier Regarding the `MDMA_APP_IDENTIFIER`… now that we are extracting the metadata as part of the creation of a Build, we can change the code not to do it again inside `PackageAnalyzer`.

In other words, we are currently passing the entire `build` object to generate the manifest:

https://github.com/clutter/mdma/blob/d584f91937334f185d7605db5c1b89771157a6ff/app/models/build.rb#L28-L30

but since are already extracting the metadata in the model, we can just:

- store the metadata inside `build.package.blob.metadata` in the model
- pass that metadata to the job to create a file

**Comment by claudioclutter** [comment]
@jeremygrenier And one more thing! Since we are now enforcing the presence and validation of the `package` *before* we make a build, then in principle we could even **delete** the `version` field from the builds table.

The reason is that we would already have this value for a build inside `build.package.metadata["bundle_version"]` because of this code

https://github.com/clutter/mdma/blob/d584f91937334f185d7605db5c1b89771157a6ff/app/jobs/generate_manifest_job.rb#L17-L19

All we need is to move that code to the model itself.

**Pull request merged**